docs/usage-guide/topics/ch04-connection.md

```diff
@@ -1,6 +1,8 @@
 # TLS Connections
 
-Users will need to create a `s2n_connection` struct to store all of the state necessary for a TLS connection. Call `s2n_connection_new()` to create a new server or client connection. Call `s2n_connection_free()` to free the memory allocated for this struct when no longer needed.
+Users interact with s2n-tls primarily through the `s2n_connection` and `s2n_config` structures. Users should build a `s2n_config` object and `s2n_connection` objects, associate the config with those connections, and then calling `s2n_negotiate()` on those connections until the TLS handshakes are completed. Users should then call `s2n_send`/`s2n_recv` on those connections to send and receive application data.
+
```

Comment:
> I'm not really sure what the goal of this paragraph is or that its inclusion makes the documentation better.

The initial intent of this paragraph is to provide a general overview of S2N software architecture. I think such a high level overview does make the documentation better, since it provides a general idea of how S2N-TLS usually works. It also connects the following chapters together. This is also suggested by [this comment](https://github.com/aws/s2n-tls/pull/4868#discussion_r1831531265).

> technically you don't have to create a config to use a connection.

I would change the wording as `In general, users build a s2n_config ...` to indicate how S2N is generally configured. That change implies there can be exceptions.

> I'm also not sure how useful the very brief mention of negotiate and send/receive is-- if we're going to mention them, we should probably at least link to the usage guide section that goes into more details.

I think the process of negotiation and send/receive is a part of how users would typically interact with S2N-TLS. I would keep this part. I agree that we need to link chapter 7 - sending and receiving to this part.

Reply:
> You may need to think about reorganizing things. Introduce what a connection is, then direct customers to the config section for configuration options and the IO section for how to use IO.

I want to set up a subsection under `# TLS Connections` called `## Software Architecture`, and move the paragraph into that subsection. Hence, the `s2n_connection` introduction will go before the general overview of software architecture.

Moreover, `## Connection Memory` subsection talks about freeing memory after handshake.
https://github.com/aws/s2n-tls/blob/01e90bb34829aa2d63c2c74534e33474085c3422/docs/usage-guide/topics/ch04-connection.md?plain=1#L9
Hence, introducing the software architecture after `# TLS Connections` and right before `## Connection Memory` should helps the doc to explain when the sequence of events usually take place.

Reply:
I'm not convinced that "Software Architecture" really makes sense as its own section. There's not much to talk about other than connections with software architecture :P

I'd go for:
1. what is a connection
2. how is a connection used, high level with links to details
3. how do connections interact with configs, high level with links to details

Reply:
I wrote a general introduction about TLS connections and configs, which includes answers to those questions. We can revise it based on the new commit.